### PR TITLE
Refine UA and balance point calculations through outlier detection

### LINKED ISSUE #13: Determine how to automatically identify outliers
Using statistics, determine a way to identify and remove outliers from UA values.  This will improve the precision of the UA value.

If the standard deviation % of UA is above a threshold (10% ?), try removing periods that are outliers in terms of UA until the % falls below the threshold.

### LINKED ISSUE #49: Make sure balance point is always less than or equal to the thermostat set point
Currently, the while loop in recalculate_bps() does not have anything preventing it from returning a balance point that's above the thermostat set point.

### LINKED ISSUE #50: refactor rules engine around a BillingPeriod class rather than using disconnected lists
The engine should work with lists of BillingPeriod objects which bundle together properties like start date, end date, usage amount, and total HDDs so that we don't have to have separate lists for each piece of data.

### PR DESCRIPTION
**Pull request by mciethan**
fixes #13 
fixes #49 
fixes #50 

To do: 
- [x] add outlier removal functionality
- [x] make sure requirements.txt reflects the addition of numpy to pyproject.toml
- [x] raise some kind of error if balance point goes as high as thermostat set point (meaning we can't produce a valid estimate for that home)
- [x] add a parameter for the minimum reduction in stdev percentage that makes the removal of an outlier worth it
- [x] consider making a BillingPeriod class which bundles together start date, end date, usage, etc
- [x] add some more tests

### CODE REVIEW

**Review by jkwan2011 — APPROVED** [review]
Looks good, thanks for updating the tests!

File: rules-engine/src/rules_engine/engine.py [hunk 1]
```diff
@@ -71,73 +49,137 @@ def average_indoor_temp(
...
+    ) -> None:
+        """Calculates the estimated balance point and UA coefficient for the home,
+        removing UA outliers based on a normalized standard deviation threshold.
+        """
+        self.uas = [bp.ua for bp in self.bills]
+        self.avg_ua = sts.mean(self.uas)
+        self.stdev_pct = sts.pstdev(self.uas) / self.avg_ua
+        self.refine_balance_point(balance_point_sensitivity)
+
+        while self.stdev_pct > 0.10:
```
Comment (jkwan2011) [inline rules-engine/src/rules_engine/engine.py:103]: Should we make these sensitivities/magic numbers configurable constants?

**Comment by mciethan** [comment]
Made 0.10 into a parameter

File: rules-engine/src/rules_engine/engine.py (hunk 1 above)
Comment (mciethan) [reply rules-engine/src/rules_engine/engine.py]: Latest commit makes 0.1 into a parameter

File: rules-engine/tests/rules_engine/test_engine.py [hunk 2]
```diff
@@ -41,23 +31,50 @@ def test_average_indoor_temp():
...
     assert ua_1 == approx(1450.5, abs=1)
     assert ua_2 == approx(1615.3, abs=1)
     assert ua_3 == approx(1479.6, abs=1)
-    assert avg_ua == approx(1515.1, abs=1)
-    assert stdev_pct == approx(0.0474, abs=0.01)
+    assert home.avg_ua == approx(1515.1, abs=1)
+    assert home.stdev_pct == approx(0.0474, abs=0.01)
 
 
 if __name__ == "__main__":
```
Comment (alanisaac) [inline rules-engine/tests/rules_engine/test_engine.py:75]: `nit`: what is this being used for?  If it's for running tests, I don't think we need this given `pytest` is our test runner.  e.g. I can delete this section and run `pytest` from my `venv` and all tests are run.  Or is it used for something else?

**Review by alanisaac — APPROVED** [review]

**Pull request merged**